### DISCUSSION TITLE
Weird behaviour with atool

### DISCUSSION THREAD

**Discussion by oredaze** [gokcehan/lf#1181] [Q&A] [answered]
New lf user here. I use atool to deal with archives.
I set a hotkey in lf: `map ac push $apack<space><space>$fx<left><left><left><left>`. A bit ugly, but it allows me to start typing a filename immediately and compress my files. Maximum speed. But when I do they receive a directory path starting all the way from `home/user/...` dir. That does not happen when I use `apack test.tar.gz file1` for example. And it shouldn't happen because I did not tell it to have folders. What am I missing.

Also I am not super versed in writing scripts, so I can only do simple solutions like this one, which are perfectly fine anyway.
EDIT:
Here is a slight improvement: `map ac push $apack<space><space>$fx<a-b><c-b><c-b>`
Also I noticed that the default config provides custom commands to archive, but it always creates a containing folder, even for 1 file...

I hope to reach a solution where it only creates a folder (one folder...) automatically only when you compress more than 1 file.

**Comment by joelim-work** [comment, marked as answer]
The exported `$fx` variable is an _absolute path_ in the form `/home/user/dir/...`. This is stated in the documentation: 

https://github.com/gokcehan/lf/blob/6ffc4a4c8aef3cf84b04b44515240612865a4e4e/doc.go#L898-L908

Most likely what happens is that when you run the archiving program, the paths specify not only the location of the files to be archived, but also its location _within_ the archive. In addition the leading `/` is removed to prevent the files from being extracted into the root directory, as a security measure.

The example given in the default config file works around the problem of absolute paths by copying the files into a temporary directory and pointing the archiving program to that directory, which is a relative path. However this does mean that the files will be placed in a containing directory as you have mentioned.

```
cmd tar ${{
    set -f
    mkdir $1
    cp -r $fx $1
    tar czf $1.tar.gz $1
    rm -rf $1
}}
```

If you don't want the containing directory, you can instruct `tar` to change to the directory before archiving the files.

```
cmd tar ${{
    set -f
    mkdir $1
    cp -r $fx $1
    tar czf $1.tar.gz -C $1 .
    rm -rf $1
}}
```

Or alternatively you can use `cd` to change to the directory. In this case a glob is used so `set -f` shouldn't be specified.

```
cmd tar ${{
    mkdir $1
    cp -r $fx $1
    cd $1
    tar czf ../$1.tar.gz *
    cd ..
    rm -rf $1
}}
```

**Comment by oredaze** [comment]
I understand now. Thank you.
I did read the documentation, but I guess there was so much to learn that I did not pay attention to this detail.
Here is my implementation, if people want to take ideas:
```
cmd tar ${{
    set -f
    if [ $(echo "$fx" | wc -l) -eq 1 ]; then
        tar czf $1.tar.gz $(basename $f)
    else
        mkdir $1
        mv $fs $1
        tar czvf $1.tar.gz $1
        mv $1/* .
        rm -d $1
    fi
}}
```
First check for a single file, if so, strip the path that LF creates and tar it. If multiple files then ensure a containing dir by moving them, tar and move them back. The reason for mv instead of cp is, as far as I understand, that will produce less disk writes than copying, since it doesn't actually change the file position on disk or create new files. I also added a -v option, in case you tar big things that take a while.
Tell me if there is any problem with that.
There are places for improvement, like the file name can't be an already existing name, but that is minor.

**Comment by mikeri** [comment]
I had the same problem and arrived here when searching for a solution. Personally I'm not a fan of creating a temporary work directory, (especially troublesome if you can't write to the current directory) so here is my variant that strips $PWD from the file paths in $fx, in case future readers might benefit from it:
```
cmd compress %{{
    name="$1"
    if [ -z "$name" ]; then
        name="$f".tar.gz
    fi

    set --
    for path in $fx; do
        rel=${path#"$PWD"/}
        set -- "$@" "$rel"
    done

    apack -a "$name" -- "$@" && echo "Created $name"
}}
```
This uses apack from atool, so you can specify the prefix of the format you want, for example ``:compress /tmp/files.zip``.
